### DISCUSSION TITLE
How to setup bash completion for `chezmoi git`?

### DISCUSSION THREAD

**Discussion by Lockszmith-GH** [twpayne/chezmoi#2408] [Q&A] [answered]
I understand that `chezmoi git` is practically calling the `git` I have installed on my system.

I was wondering if there is a way to setting completion based on the `bash-completion` that is already setup for `git` and works successfully?

I'm also ok with defining an alias/function named `czg` and have that mapped to git, but I'm hitting a wall with that.
Has anybody successfully setup something like that?
Any suggestion here would be welcome.

**Comment by twpayne** [comment]
I don't think this is possible with current shell completion systems.

The fundamental problem is that `chezmoi git` runs `git` in the source directory and so should use git's shell completions, but shell completion systems have not been designed with composability in mind. What is needed is for chezmoi's shell completions to say "when you see `chezmoi git`, use git's shell completions instead". I do not know of any way of achieving this.

**Reply by Lockszmith-GH** [reply]
I'm not giving up hope yet, I've gone in a similar path - and I'm ok with `chezmoi git` or even an alias (let's say `czg`) only auto-completing `git` commands vs full contextual namespaces.
If I ever find a hack/solution, I'll update here.

**Reply by twpayne** [reply]
An alternative option that might work for you is the command `chezmoi git` is similar to `git --git-dir=$HOME/.local/share/chezmoi/.git`, i.e.

```bash
alias czg="git --git-dir=$HOME/.local/share/chezmoi/.git"
```

might be part of a solution.

**Reply by Lockszmith-GH** [reply, marked as answer]
This was a great idea, it does however require an additional step. See [these](https://stackoverflow.com/q/11969133/799379) [two](https://superuser.com/q/436314/100730) for further insight.

At this point, the following steps makes my `czg` alias working for me exactly as expected:

### Step # 1
Load the code from [complete_alias](https://github.com/cykerway/complete-alias/blob/master/complete_alias) (which I use chezmoi to keep up to date 🎉 )

### Step # 2
Define the alias (NOTE: you had `--wort-tree` parameter missing, it is required+)
```
CZG_ROOT="$(chezmoi git -- rev-parse --show-toplevel)"
alias czg="git --work-tree=$CZG_ROOT --git-dir=$CZG_ROOT/.git "
unset CZG_ROOT
```

### Step # 3
Use the complete_alias provided method to create the completion code:
```
complete -F _complete_alias czg 
```

And now tab completion works!

(I told you I will not lose hope 😄)

**Reply by twpayne** [reply]
Woo-hoo! Nicely done! I must say I am very pleasantly surprised :)
